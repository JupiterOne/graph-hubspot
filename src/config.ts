import {
  IntegrationExecutionContext,
  IntegrationInstanceConfig,
  IntegrationInstanceConfigFieldMap,
  IntegrationValidationError,
} from '@jupiterone/integration-sdk-core';
import { createAPIClient } from './client';

export const instanceConfigFields: IntegrationInstanceConfigFieldMap = {
  oauthAccessToken: {
    type: 'string',
    mask: true,
  },
  apiBaseUrl: {
    type: 'string',
    mask: true,
  },
};

/**
 * Properties provided by the `IntegrationInstance.config`. This reflects the
 * same properties defined by `instanceConfigFields`.
 */
export interface IntegrationConfig extends IntegrationInstanceConfig {
  /**
   * This access_token is considered valid as a Bearer Authorization header
   */
  oauthAccessToken: string;

  /**
   * Hubspot API base url
   */
  apiBaseUrl: string;
}

export async function validateInvocation(
  context: IntegrationExecutionContext<IntegrationConfig>,
) {
  const { config } = context.instance;
  if (!Object.keys(instanceConfigFields).every((key) => config[key])) {
    throw new IntegrationValidationError(
      `Config requires all of {${Object.keys(instanceConfigFields).join(
        ', ',
      )}}`,
    );
  }

  const apiClient = createAPIClient(config, context.executionHistory);
  await apiClient.verifyAuthentication();
}
