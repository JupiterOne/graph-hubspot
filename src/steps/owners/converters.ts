import {
  createIntegrationEntity,
  parseTimePropertyValue,
} from '@jupiterone/integration-sdk-core';
import { Owner } from '../../types';
import { getEntityKey } from '../../utils';
import { Entities } from '../constants';

export function createUserEntity(data: Owner) {
  return createIntegrationEntity({
    entityData: {
      source: data,
      assign: {
        _class: Entities.USER._class,
        _type: Entities.USER._type,
        _key: getEntityKey(Entities.USER, data.id),
        id: data.id,
        name:
          data.firstName && data.lastName
            ? `${data.firstName} ${data.lastName}`
            : data.email,
        firstName: data.firstName,
        lastName: data.lastName,
        email: data.email,
        username: data.email,
        archived: data.archived,
        userId: data.userId,
        createdOn: parseTimePropertyValue(data.createdAt),
        updatedOn: parseTimePropertyValue(data.updatedAt),
      },
    },
  });
}
