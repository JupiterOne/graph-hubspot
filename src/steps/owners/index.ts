import {
  createDirectRelationship,
  IntegrationStep,
  IntegrationStepExecutionContext,
  RelationshipClass,
} from '@jupiterone/integration-sdk-core';
import { createAPIClient } from '../../client';
import { IntegrationConfig } from '../../config';
import { getEntityKey } from '../../utils';
import { Entities, IntegrationSteps, Relationships } from '../constants';
import { createUserEntity } from './converters';

export async function fetchOwners({
  instance,
  executionHistory,
  jobState,
}: IntegrationStepExecutionContext<IntegrationConfig>) {
  const apiClient = createAPIClient(instance.config, executionHistory);
  const res = await apiClient.hubspotClient.crm.owners.defaultApi.getPage();

  for (const owner of res.body.results) {
    const userEntity = createUserEntity({
      ...owner,
      createdAt: owner.createdAt.toString(),
      updatedAt: owner.updatedAt.toString(),
    });
    await jobState.addEntity(userEntity);

    if (owner.userId) {
      await apiClient.fetchUser(owner.userId.toString(), async (user) => {
        const { roleId } = user;
        if (roleId) {
          const roleEntity = await jobState.findEntity(
            getEntityKey(Entities.ROLE, user.roleId.toString()),
          );

          if (roleEntity && userEntity) {
            await jobState.addRelationship(
              createDirectRelationship({
                _class: RelationshipClass.ASSIGNED,
                from: userEntity,
                to: roleEntity,
              }),
            );
          }
        }
      });
    }
  }
}

export const ownerSteps: IntegrationStep<IntegrationConfig>[] = [
  {
    id: IntegrationSteps.OWNERS,
    name: 'Fetch Owners',
    entities: [Entities.USER],
    relationships: [Relationships.USER_ASSIGNED_ROLE],
    dependsOn: [IntegrationSteps.ROLES],
    executionHandler: fetchOwners,
  },
];
