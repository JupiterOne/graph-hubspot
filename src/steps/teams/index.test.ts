import { createMockStepExecutionContext } from '@jupiterone/integration-sdk-testing';
import * as team from '.';
import { createIntegrationConfig } from '../../../test/config';
import { IntegrationConfig } from '../../config';
import * as owner from '../owners';

describe('#fetchTeams', () => {
  // TODO: we'll add recordings later
  // let recording: Recording;

  // beforeEach(() => {
  //   recording = setupHubSpotRecording({
  //     directory: __dirname,
  //     name: 'fetchOwners',
  //   });
  // });

  // afterEach(async () => {
  //   await recording.stop();
  // });

  test('should collect data', async () => {
    const context = createMockStepExecutionContext<IntegrationConfig>({
      instanceConfig: createIntegrationConfig(),
    });

    await owner.fetchOwners(context);
    await team.fetchTeams(context);

    expect({
      numCollectedEntities: context.jobState.collectedEntities.length,
      numCollectedRelationships: context.jobState.collectedRelationships.length,
      collectedEntities: context.jobState.collectedEntities,
      collectedRelationships: context.jobState.collectedRelationships,
      encounteredTypes: context.jobState.encounteredTypes,
    }).toMatchSnapshot();

    const teams = context.jobState.collectedEntities.filter((e) =>
      e._class.includes('Team'),
    );

    expect(teams).toMatchGraphObjectSchema({
      _class: ['Team'],
      schema: {
        additionalProperties: false,
        properties: {
          _rawData: {
            type: 'array',
            items: { type: 'object' },
          },
          _type: { const: 'hubspot_team' },
          // We want to make sure we include every field from converter.ts
          name: { type: 'string' },
        },
      },
    });
  });
});
