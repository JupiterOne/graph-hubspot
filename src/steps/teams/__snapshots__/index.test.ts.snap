// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#fetchTeams should collect data 1`] = `
Object {
  "collectedEntities": Array [
    Object {
      "_class": Array [
        "Team",
      ],
      "_key": "hubspot_team:3515767",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "id": "3515767",
            "name": "Team A",
          },
        },
      ],
      "_type": "hubspot_team",
      "createdOn": undefined,
      "displayName": "Team A",
      "id": "3515767",
      "name": "Team A",
    },
  ],
  "collectedRelationships": Array [],
  "encounteredTypes": Array [
    "hubspot_team",
  ],
  "numCollectedEntities": 1,
  "numCollectedRelationships": 0,
}
`;
